Feature/custom start row

chrisjfield opened the pull request:
I have tried to implement #140 (custom start rows)

Python is not my primary language so hopefully, it's not too bad!

chrisjfield commented:
Apologies for the many commits to fix the checks - was struggling to set up the same checks locally so missed a few bits in the initial commits - looks like it all passes now :)

centreboard commented:
Thanks for doing this, looks good so far.

I have a couple of open questions around what behaviour we'd expect to happen with a custom start row:
1) Should `rounds` at the end of a touch also be the starting row or not? i.e. what should happen if you also specify `--stop-at-rounds` or after `That's all`?
I think I've mostly started practices at custom rows to ring different parts of a course, e.g. start Plain Bob Minor from `135264` to ring the second half course

2) How should this interact if you specify a composition id?
a) Look for the row in the composition (ignoring cover bells) and start from that point.
b) Transpose the composition so, using the custom row as the start instead of rounds

chrisjfield commented:
Good questions!

1. Personally I think your use case would be the overwhelming one, so I would expect the custom row at the start and rounds at the end. I would think that stop at rounds should still stop at rounds rather than the custom row. 

2. my thoughts would be to go with b), it seems to allow more flexibility and be significantly less upfront work for Wheatly, happy to be guided differently though.

I'll have a play with those bits and add something in later :)

kneasle commented on CONTRIBUTING.md:
```diff
@@ -24,22 +28,27 @@ Or, on Unix you can run `./run-wheatley [ARGS]`.
-│   ├── arg_parsing.py                              => Code to parse formatted CLI args e.g. call strings
+│   ├── aliases.py
 │   ├── bell.py                                     => Stores the representation of a bell
 │   ├── bot.py                                      => The main class for Wheatley
 │   ├── calls.py                                    => Constants for the calls used in Ringing Room
 │   ├── main.py                                     => Entry point of Wheatley
 │   ├── page_parser.py                              => Code to parse the ringing room HTML
-│   ├── regression.py                               => Code for the weighted linear regression
-│   ├── rhythm.py                                   => Code used to drive Wheatley's rhythm
+│   ├── parsing.py                                  => Code to parse formatted CLI args e.g. call strings
+│   ├── stroke.py                                   => A module containing a type that encapsulates a stroke
 │   ├── tower.py                                    => Code used to interact with ringing room
+│   ├── rhythm.py                                   => Code used to drive Wheatley's rhythm
+|   |   ├── __init__.py
+|   |   ├── abstract_rhythm.py
+|   |   ├── regression.py                           => Code for the weighted linear regression
+|   |   ├── wait_for_user.py
```
Nice! Thanks for updating this.  I'm planning to improve the documentation and this is a good start (if I'd known you were going to make a PR I'd have done it sooner, but you seem to have done well anyway).

kneasle commented on wheatley/bot.py:
```diff
@@ -137,7 +138,15 @@ def _on_row_gen_change(self, row_gen_json: JSON) -> None:
     def _on_size_change(self) -> None:
         self._check_number_of_bells()
-        self._rounds = rounds(self.number_of_bells)
+        self._check_starting_row()
+        self._rounds = generate_starting_row(self.number_of_bells, self.row_generator.custom_start_row)
```
I think these are the wrong way round - we should check the newly created starting row?

kneasle reviewed (commented):
Thanks for making this PR - it's always nice to see new names on PRs :grin:.

> Apologies for the many commits to fix the checks - was struggling to set up the same checks locally so missed a few bits in the initial commits - looks like it all passes now :)

That's OK - we've been trying to catch as many potential bugs in the PR checks, so they've ended up being quite strict.  Also, none of the checking programs ship with Python, so we should probably add it to `CONTRIBUTING.md` (though I think they're in `requirements.txt`, though - @centreboard?).  I also think it would be good to make a shell script to run all the PR checks, but this means we'd have to keep the CI and this script in sync.

kneasle commented:
So my thoughts on how this should behave:
- I agree with @centreboard that starting a method or dixonoid from a custom start row should still stop at true rounds if `--stop-at-rounds` is set or the user calls `That's All` (anything else is, I think, unintuitive).
- I'm not sure that allowing `--start-row` in compositions is a good idea, and think it should simply cause an error.  We basically have 3 options to handle this:
  - Search for the first occurrence of the `--start-row`, and start from there.  This is not well defined for multi-extent blocks (which row do we start at), or sub-extent compositions (what if the row doesn't exist), and also risks starting at a completely random point in the lead of any method.
  - Transpose the entire composition by `--start-row`.  I think this is kinda weird if the composition is defined in terms of calling positions, since starting at a random point could result in a wildly different calling.  Also, the behaviour of `--stop-at-rounds` isn't well defined.
  - Throw an error when Wheatley starts up.  Not ideal, but it is well defined and it's obvious to the user that this won't work.
- What should be the behaviour of changing the cover bell using `--start-row`?  My gut reaction is that things should just work, and the last bell in `--start-row` is the cover bell.
- I'm not sure what `That's All` should do - should it return Wheatley to rounds (as suggested) or `--start-row` (to restart the touch).

This is not a prescription, just my thoughts - I'm open to other ideas and discussion is always good.

chrisjfield commented:
> That's OK - we've been trying to catch as many potential bugs in the PR checks
Very sensible too! I think my main issue was I have prettier set up to auto-format/fix in VS code and it was sneakily changing stuff! I have turned off for python for now.

> I think these are the wrong way round - we should check the newly created starting row?
Good catch, should be fixed now

> should still stop at true rounds if --stop-at-rounds
I agree, that should be resolved in the latest commit

> I'm not sure what That's All should do
I have changed it to ring "normal" rounds currently, think there is an argument both ways but this way felt better to me.
You can always stand and go again to quickly return to the custom row, but if it's set the other way it's hard to return to rounds without restarting Wheatley. My logic at least!

> What should be the behaviour of changing the cover bell using --start-row? 
Checked this out and it works as I would expect (the last bell is cover) - I think that makes sense

>I'm not sure that allowing --start-row in compositions is a good idea
I was working on it and came to the same conclusion, it just doesn't really make sense in a pre-defined composition to then randomly change the start row. Also causes issues on recognising the end etc, I have just added a catch for it and stop Wheatly running up.

I was trying to extend the mutually exclusive arguments but can't seem to find any evidence that argparser supports overlapping exclusive groups - so there may well be a neater solution than my route one if clause!

centreboard commented on wheatley/bot.py:
```diff
@@ -137,7 +140,17 @@ def _on_row_gen_change(self, row_gen_json: JSON) -> None:
         self._rounds = rounds(self.number_of_bells)
+        self._check_starting_row()
+
+    def _check_starting_row(self) -> bool:
+        if len(self._opening_rounds) != self.number_of_bells:
+            self.logger.warning(f"Row generation requires {self.row_generator.stage} "
+                                + "bells in the starting row Wheatley will not ring!")
+            return False
+        return True
```
I think the error message should refer to `self._opening_rounds` as that's the given part. Something like
```suggestion
    def _check_starting_row(self) -> bool:
        if len(self._opening_rounds) > self.number_of_bells:
            self.logger.warning(f"The current tower has fewer bells ({self._tower.number_of_bells})"
            + "than the starting row {self._opening_rounds}. Wheatley will not ring!")
            return False
        return True
```

Using `len(self._opening_rounds) > self.number_of_bells` as the not ringing condition would allow `13524` as a starting row for Plain Bob Doubles, but it doesn't save much over typing the cover bell.

chrisjfield commented on wheatley/row_generation/helpers.py:
```diff
@@ -74,3 +74,11 @@ def convert_to_bell_string(bell: int) -> str:
+    if start_row is None:
+        return rounds(number_of_bells)
+
+    return Row([Bell.from_str(i) for i in start_row])
```
I'm not sure it has to contain every number once.
On init RowGenerator only knows about the stage and not the number of bells if I am reading it right.
Say you ring Doubles but want the 4 as cover, you might enter "12356" so checking all bells in the range 1-5 are used would break incorrectly.

I can check that the bells are unique and add any missing bells in to the end of the change.

The only oddity I have found doing that is if you ring doubles and enter a starting row of "4321" we will expand that to "432156" which is fine, but we are adding a working bell in, not sure if we want that or would rather error and force a user to enter it

kneasle replied:
I agree with @centreboard here - we should verify that the input row is valid along with the other args.  Currently you could input `--start-row 414` and nothing would break (except that you'd get shedloads of 'bell on wrong stroke' errors because the 4 is rung twice in every row).  Also, I suspect that in the vast majority of cases invalid rows are simply a mistake on behalf of the user so it would be nice to throw an upfront error (anyway rows have to contain each bell exactly once in order to be classified as change ringing).  I implemented checks like this for a separate project and it works nicely - [here](https://github.com/kneasle/3rd-yr-proj/blob/master/core/src/row.rs#L407-L430)'s the code in that project, which might be helpful (it's in Rust though which is not very widely used).

So, what I suggest is that the incoming `--start-row` arg is parsed and verified along with the other args (e.g. call strings, peal speed, etc.) - this should probably add another function to `parsing.py` with appropriate error classes for missing and duplicated bells.  We don't verify the stage at this point, because that could change whilst Wheatley is running.  Then we pass the parsed and verified row into `Bot.__init__` and then everything else is already done.

centreboard replied:
@chrisjfield is right we could just append missing bells onto the end. My suspicion is entering `13256` for doubles is more likely to be accidentally not typing `4` than wanting it as a cover bell. If Wheatley adds it, there should probably be an information or warning logged.
If we expand to add cover bell, I'm also fine with expanding to add working bells (so `1423` could work for minor).

The short form I'd probably use most want to use is `14325` for doubles on 5, 6 or more bells.

Ultimately these are really edge cases and I'm happy with whatever's cleanest to code such that Wheatley ends up with valid rows, and we've communicated validation errors back to users.

kneasle replied:
Hmmm reading my comments I haven't been very clear... so my proposed idea is the following:
- The user enters some string (say `--start-row 1423`)
- Wheatley parses this string into a row of the smallest possible stage while still in the `main` function, in this case `[1, 4, 2, 3]`.  If this row is invalid, then we should explain the error somehow and terminate, e.g.:
  - For `1442` we should emit something like `Bell '4' is duplicated`
  - For `124` we either emit `Bell '3' is missing` or `Bell '4' is too big`.  Both are equally valid but the former is probably more intuitive.
- This small but valid row is then passed into `Bot` as the starting row, and has consecutive bells appended to the end to extend it to the correct number of bels (just like how we currently add cover bells).  This process can't invalidate a row, so it's fine to do without worries about misinterpretation.  If the unextended row already has too many bells, then we emit a warning and crash on entering changes (similar to ringing e.g. major on 6 bells).

This way, I think we get the best of all worlds - Wheatley helpfully reports potential errors and can't ring any invalid rows, but we also get the nice ability to permute the little bells without typing all the rounds.  It also has the nice side effect that `""` parses to the empty row `[]` which then expands to rounds on any stage, so we don't have to handle rounds as a special case.

chrisjfield replied:
I was playing with this last night and have something locally that works (I think) but was waiting to see what your thoughts were. It seems similar to what you suggest above but with a slight difference so interested in what you think.

What I have at the moment is:
* The user enters some string (say --start-row 1423)
* On init we parse the string using the stage
** if duplicate bells are received we throw a value error
** if bells are missing we append to the end in order (so -- start-row 134 with stage 5 becomes 13425)
* In bot we re-generate using the actual number of bells in the tower, so if there are 6 bells it now becomes 134256
* Before we go into changes we do some extra checks:
** check the inputted starting row is equal or shorter to the number of bells in the tower (just to allow nice messaging)
** check the actual starting row equals the number of bells in the tower.

I think this is similar but allows for inputting shorter starting rows where you want unusual cover bells. This is obviously an edge edge case but may as well let it go IMO.

It also indirectly catches out of range errors as they would end up with a starting row longer than the number of bells.
I think it's best to check here as the row might not start valid but could become valid when the tower changes size etc so seems odd to error before they have a chance to fix it.

The only oddity I found is it means Wheatly can add bells which are working bells. As @centreboard mentioned that might be ok. If we don't like it it's easy enough to add another validation that the inputted row is >= to the stage in length.

Hope that makes sense, thoughts welcome!

kneasle commented:
> Good catch, should be fixed now

Cool.  Btw, once a conversation or review point has been dealt with, the convention is to click the 'Resolve conversation' button at the bottom and it will be marked as done and collapsed.  This way, only outstanding things are shown and it's much more readable. I usually leave a comment in that convo too but that's optional.

> I have changed it to ring "normal" rounds currently

Yep I think your logic is sound - this feels most intuitive to me and is also consistent with `--stop-at-rounds`.

> Checked this out and it works as I would expect (the last bell is cover) - I think that makes sense

Noice!

> I was trying to extend the mutually exclusive arguments but can't seem to find any evidence that argparser supports overlapping exclusive groups - so there may well be a neater solution than my route one if clause!

Yeah I don't think this is possible.  I think just a simple check and error in `console_main()` would suffice (I haven't looked at the code recently so you may have done this).

centreboard commented on wheatley/bot.py:
```diff
@@ -81,8 +82,10 @@ def __init__(self, tower: RingingRoomTower, row_generator: RowGenerator, do_up_d
         self._place = 0
-        self._rounds: Row = rounds(MAX_BELL)
+        self._opening_rounds: Row = row_generator.start_row
+        self._rounds: Row = rounds(self.number_of_bells)
```
Is `self.number_of_bells` well defined at this point? I have a feeling `_on_size_change` is called once the tower is connected and joined the tower, so this is more a place holder

chrisjfield replied:
your right, I call generate_starting_row in _on_size_change for that reason. 
I'm happy to switch back to rounds(MAX_BELL) not sure it matters too much?

centreboard reviewed (commented):
Looking really good. 
Looks like the Dixon's tests pass in the default plain rules as a positional argument

chrisjfield commented on wheatley/row_generation/helpers.py:
```diff
@@ -74,3 +74,21 @@ def convert_to_bell_string(bell: int) -> str:
+
+    # If there are more bells than given in the starting row
+    # add the missing ones in sequential order as cover bells
+    for i in range(1, number_of_bells):
```
This gets re-called from _on_size_change when it starts rining which passes in the true number of bells so I think it will always end up correct.

centreboard replied:
A failing test case would be

```
self.assertListEqual(Row([Bell(1), Bell(0), Bell(2), Bell(3)]), generate_starting_row(4, "2"))
```

Range isn't inclusive, so this wouldn't add the tenor.

chrisjfield replied:
ah, thanks, my python knowledge letting me down!
really should have had a test for that, added it now

chrisjfield commented on CONTRIBUTING.md:
````diff
@@ -1,17 +1,39 @@
+```
+py -m pytest
+```
+
````
I added this in mainly for me so I don't forget parts of this when testing!

centreboard commented on wheatley/bot.py:
```diff
@@ -137,7 +140,23 @@ def _on_row_gen_change(self, row_gen_json: JSON) -> None:
+    def _check_starting_row(self) -> bool:
+        if (self.row_generator.custom_start_row is not None and
+                len(self.row_generator.custom_start_row) < self.number_of_bells):
+            self.logger.info("The starting row contains fewer bells than the tower"
+                             + f"({self._tower.number_of_bells}) "
+                             + "Wheatly will add the extra bells to the end of the change")
```
```suggestion
            self.logger.info(f"The starting row '{self.row_generator.custom_start_row}' "
                             + f"contains fewer bells than the tower ({self._tower.number_of_bells}). "
                             + "Wheatley will add the extra bells to the end of the change.")
```

centreboard replied:
In _check_number_of_bells line 171 should check if there's a custom row

```py
if self._tower.number_of_bells > self.row_generator.stage + 1 and self.row_generator.custom_start_row is None:
```
Cover bells are added to `self._opening_rounds` by construction, with a message logged there when `custom_start_row` is set.

Currently the line 171 check incorrectly logs `-m "Plain Bob Doubles" --start-row 13572468`
> INFO:BOT:Current tower has more bells (8) than expected (6). Wheatley will add extra cover bells.

chrisjfield replied:
Good catch

centreboard commented on wheatley/row_generation/helpers.py:
```diff
@@ -74,3 +74,21 @@ def convert_to_bell_string(bell: int) -> str:
+
+    # If there are more bells than given in the starting row
+    # add the missing ones in sequential order as cover bells
+    for i in range(1, number_of_bells + 1):
```
Currently if you give this a bigger bell (e.g. `generate_starting_row(4, "T")` it generates `T1234` instead of `T1234567890E`

```py
self.assertListEqual(Row([Bell(5), Bell(0), Bell(1), Bell(2), Bell(3), Bell(4)]), generate_starting_row(4, "6"))
```

Needs something like
```py
    highest_given_bell = max(bell.number for bell in custom_start_row)
    upper_limit = max(number_of_bells, highest_given_bell) + 1
    for i in range(1, upper_limit):
```

chrisjfield replied:
This one was on purpose, my thinking was that number_of_bells is the key thing, we should never have bells that are higher than that.

So there are 2 cases I think.
We start with some case (your example) `generate_starting_row(4, "6")` => `64321`

Then either:
* In bot it turns out there are actually 6 bells in the tower, this then gets rerun as `generate_starting_row(4, "6")` => `654321`
* There are actually only 4 (or any number fewer than 6) - then in _check_starting_row the validation `len(self._opening_rounds) != self.number_of_bells` will fail and tell the user that the starting row is too long.

I was thinking the most common case is just a mistake where you specify a number too high.
I grant the error is not the best, but it also catches where you just enter a string that's too long and I was not sure of an easy way to separate the validations

centreboard replied:
An advantage of padding out to the max bell would be it always returns a valid row. Checking `len(self._opening_rounds) != self.number_of_bells` would still mean if the entered string contained `T` then the tower would need at least 12 bells.

There are 2 places this method is called, once in row_generator and again in the bot. The bot updates opening rounds as tower size changes, row_generator doesn't need to update once it is created. 

Currently the subtle bug I hit was is with ` -m "Plain Hunt on 4" --start-row 6` in a 6 bell tower
> INFO:BOT:ROW: 6 1 2 3 4 5
INFO:BOT:Starting to ring Plain Hunt on 4
INFO:BOT:ROW: 6 1 2 3 4 5
INFO:BOT:ROW: 1 6 3 2 4 6
ERROR:TOWER:Bell 6 on opposite stroke

If you add `--verbose` you get
> DEBUG:ROWGEN:Reset
DEBUG:ROWGEN:[1, 6, 3, 2, 4]
INFO:BOT:ROW: 1 6 3 2 4 6

row_generator calls `generate_starting_row(4, "6")` and gets back a 5 bell list `61234` for the starting row.
It uses this row and permutes it with the methods place notation up to its stage (i.e. it preserves cover bells it knows about)
As the 5 bell row returned is less than the 6 bells in the tower, cover bells are added from rounds
https://github.com/kneasle/wheatley/blob/5eb2447c1ea6fb170d001615939f98e36fd37aeb/wheatley/bot.py#L275-L278

- Line 278 should now add cover bells from `self.opening_rounds`

kneasle replied:
OK right I haven't checked in on this for a while - I was working on #162 last night and didn't have time to look at this.  Generally, things seem to be going really well!

But I think that the weird bugs that are being discovered are more evidence that throwing a hard error while in the `main` function is the correct response to the user entering an invalid row (`143` is invalid because it doesn't contain `2` even if Wheatley tries to coerce the user's inputs to be valid).  I just think that most people wouldn't expect this behaviour, and ultimately I will go with whatever Wheatley's users think is most intuitive (also very few people actually read any documentation - I know this from how many people have asked questions which are answered in the docs) - to that end I'm tempted to make a poll on Wheatley Whisperers or the main ringing group to see what people think. @centreboard @chrisjfield what do you think to that idea?

It's also consistent with how all the other CLI args work (they all have quite strict parsing rules to reject ambiguous or incorrect inputs).  It also wouldn't be too hard to make this change (I think the resulting code might even be simpler).

chrisjfield replied:
I think thats ok.

The only downside is if you want a non-standard covering bell you have to specify it manually - but not sure thats a huge issue really.

chrisjfield replied:
Added in a parser and some tests on it, should now force you to enter a string containing all valid bells between 1 and the highest bell specified (i.e. the length of the string)

kneasle replied:
Awesome!  I'll checkout this branch on local and have a play around.

chrisjfield commented:
Is there an easy way to run the integration tests locally? 
Thought I had all the tests down 🤣.

kneasle commented:
> Is there an easy way to run the integration tests locally?
> Thought I had all the tests down :rofl:.

I think running `pytest` in the project root should do it.

> Also on the note of tests, when running the linter, does everyone get loads of "unexpected-line-ending-format" lints or is that just my VS code?

Oh you're running Windows, right?  This usually happens because git is made by Linus Torvalds (as in Linux) so git accepts only unix-style line endings as valid, whereas Windows line endings add a carriage return which makes them one character longer.  There should be a setting on VS code to set the line ending type.

centreboard commented:
> Also on the note of tests, when running the linter, does everyone get loads of "unexpected-line-ending-format" lints or is that just my VS code?
> 
> I have disabled the rule locally but didn't want to push that as I was guessing it was my set up?

It's git doing auto CRLF conversion on Windows. When it checkouts out files git converts the line endings to CRLF, when you commit it converts back to LF.

I think we could change
https://github.com/kneasle/wheatley/blob/5eb2447c1ea6fb170d001615939f98e36fd37aeb/.gitattributes#L1-L2 to be `* eol=LF` but need to work out how to test that. 

Really it's a linter warning that's of use only on the server due to the conversion being automatic on Windows, and unlikely to occur on Linux.

centreboard commented:
> Is there an easy way to run the integration tests locally?
Thought I had all the tests down 🤣.

Good question, I'm not sure you can run the doc tests (`python .\doctests`) on Windows other than using WSL at the moment. #163

`python .\fuzz` works on Windows, but that's unlikely to be needed here.

centreboard reviewed (approved):
Great work. 1 nit on spelling of Wheatley

centreboard commented on wheatley/bot.py:
```diff
@@ -137,7 +140,23 @@ def _on_row_gen_change(self, row_gen_json: JSON) -> None:
         self._rounds = rounds(self.number_of_bells)
+        self._check_starting_row()
+
+    def _check_starting_row(self) -> bool:
+        if (self.row_generator.custom_start_row is not None and
+                len(self.row_generator.custom_start_row) < self.number_of_bells):
+            self.logger.info("The starting row contains fewer bells than the tower"
+                             + f"({self._tower.number_of_bells}) "
+                             + "Wheatly will add the extra bells to the end of the change")
```
```suggestion
            self.logger.info(f"The starting row '{self.row_generator.custom_start_row}' "
                             + f"contains fewer bells than the tower ({self._tower.number_of_bells}). "
                             + "Wheatley will add the extra bells to the end of the change.")
```

centreboard replied:
```suggestion
            self.logger.info(f"The starting row '{self.row_generator.custom_start_row}' "
                             + f"contains fewer bells than the tower ({self._tower.number_of_bells}). "
                             + "Wheatley will add the extra bells to the end of the change.")
```

kneasle commented on tests/row_generation/test_Helpers.py:
```diff
@@ -71,6 +72,19 @@ def test_convert_pn(self):
+        self.assertListEqual(Row([Bell(0), Bell(1), Bell(2), Bell(3)]), generate_starting_row(4))
+
+    def test_generate_with_custom_row(self):
+        self.assertListEqual(Row([Bell(3), Bell(2), Bell(1), Bell(0)]), generate_starting_row(4, "4321"))
+        self.assertListEqual(Row([Bell(3), Bell(2), Bell(1), Bell(0)]), generate_starting_row(4, "432"))
+        self.assertListEqual(Row([Bell(3), Bell(1), Bell(0), Bell(2)]), generate_starting_row(4, "42"))
+        self.assertListEqual(Row([Bell(1), Bell(0), Bell(2), Bell(3)]), generate_starting_row(4, "2"))
```
Shouldn't these tests break if the row parsing throws an error on incorrect rows?

kneasle commented on tests/row_generation/test_Helpers.py:
```diff
@@ -71,6 +72,19 @@ def test_convert_pn(self):
+        self.assertListEqual(Row([Bell(0), Bell(1), Bell(2), Bell(3)]), generate_starting_row(4))
+
+    def test_generate_with_custom_row(self):
+        self.assertListEqual(Row([Bell(3), Bell(2), Bell(1), Bell(0)]), generate_starting_row(4, "4321"))
+        self.assertListEqual(Row([Bell(3), Bell(2), Bell(1), Bell(0)]), generate_starting_row(4, "432"))
+        self.assertListEqual(Row([Bell(3), Bell(1), Bell(0), Bell(2)]), generate_starting_row(4, "42"))
+        self.assertListEqual(Row([Bell(1), Bell(0), Bell(2), Bell(3)]), generate_starting_row(4, "2"))
```
Shouldn't these tests break if the row parsing throws an error on incorrect rows?

centreboard replied:
The parser is checking for that, so the check doesn't need to be duplicated in `generate_starting_row`.
These test cases should probably be valid strings now (except maybe 1 that would be stopped by parser)

kneasle replied:
Ah OK that makes sense

kneasle commented on wheatley/bot.py:
```diff
@@ -149,7 +168,8 @@ def _check_number_of_bells(self) -> bool:
                                 + "Wheatley will not ring!")
             return False
-        if self._tower.number_of_bells > self.row_generator.stage + 1:
+        if (self._tower.number_of_bells > self.row_generator.stage + 1
+                and self.row_generator.custom_start_row is None):
```
Bit of a nitpick, but it's probably clearer to finish with a `\` instead of using brackets:
```python
        if self._tower.number_of_bells > self.row_generator.stage + 1 \
                and self.row_generator.custom_start_row is None:
```

chrisjfield replied:
Cool, I picked this up for some python learning so useful tip :)

kneasle commented on wheatley/bot.py:
```diff
@@ -137,7 +140,23 @@ def _on_row_gen_change(self, row_gen_json: JSON) -> None:
 
     def _on_size_change(self) -> None:
         self._check_number_of_bells()
+        self._opening_rounds = generate_starting_row(
```
I think it would be clearer if `self._opening_rounds` were renamed to `self._opening_row`, since it isn't necessarily actually rounds.

kneasle commented on wheatley/main.py:
```diff
@@ -405,6 +414,11 @@ def console_main(override_args: Optional[List[str]], stop_on_join_tower: bool) -
         sys.exit(f"Bad value for '--url': {e}")
 
+    try:
+        parse_start_row(args.start_row)
```
I think that we should fully _parse_ the row here rather than just verifying it, because by only passing around strings (which are not inherently valid) we run the risk of accidentally passing an invalid row into `Bot`.

So we'd want something like:
```python
try:
    starting_row: Row = parse_start_row(args.start_row)
except:
   ...
```

In fact, I think it would make sense to move this into `create_row_generator` so that we only have to pass the unparsed args into it.

kneasle commented on wheatley/parsing.py:
```diff
@@ -6,11 +6,53 @@
+    def __str__(self) -> str:
+        return f"Error parsing start row '{self.start_row_string}': {self.message}"
+
+def parse_start_row(start_row: str) -> int:
```
I think this should return an entirely parsed `Row` - the currently returned value will be `len(row)`.  Also, the contents of `generate_starting_row` should be combined with this function because if we return a `Row` then there's nothing to generate.

centreboard replied:
We couldn't use a row returned from this directly, as it could be extended by either row gen or bot if they need more bells.
That in turn has a risk of manipulating the same list object.

chrisjfield replied:
Yeah, I agree with @centreboard, this covers a few of your comments.

The issue with fully parsing and passing in a row is that we don't actually know how many bells there are.
We need to re-generate when we actually start ringing (or tower size changes) and to do that you need the original string (or some mutable object but that seems worse)

Because of that I didn't see the point in generating the row just to add an extra property that would not be used in the end.

kneasle commented on wheatley/parsing.py:
```diff
@@ -6,11 +6,53 @@
+        max_bell = max(max_bell, bell.number)
+
+    if max_bell > len(start_row):
+        exit_with_message(f"Start row does not contain all bells from 1 to {max_bell}")
```
I think this error message would be nicer if it were more specific and told us which bell(s) are missing - even just outputting the smallest missing bell would be better.

I have some more test cases for row parsing in [a different project](https://github.com/kneasle/3rd-yr-proj/blob/master/core/src/row.rs#L793-L858) which give an example of nice error messages (they are written in Rust though, so I'm not sure how readable they are).

chrisjfield replied:
I have extended this to give the list of missing bells.
Seeing as at that point it's basically free I also added a check for dupe bells at this point (which can name the exact bell).

This now means the dupe bell error in helper should be unreachable, so hopefully, that helps with the comment below on that error.

kneasle commented on wheatley/row_generation/complib_composition_generator.py:
```diff
@@ -188,7 +188,7 @@ def summary_string(self) -> str:
         if index < len(self.loaded_rows):
             return self.loaded_rows[index]
-        return (self.rounds(), [])
+        return (self.start_row, [])
```
Is this change actually needed, since `start_row` should always be rounds for `ComplibCompositionGenerator`.  I wonder if it would be better to add a sanity check here to make sure that the code doesn't break silently.  Something like:
```python
assert self.start_row == self.rounds()
return (self.start_row, [])
```

centreboard replied:
I'm not sold on the benefit of the assert over directly `return (rounds(self.stage), [])`

chrisjfield replied:
Agreed, I originally took the rounds() property off of row_generator but that was before the original thread around compositions.

I have just added it back and returned this to it's original state - seeing as we always want to start from rounds anyway

kneasle commented on wheatley/row_generation/helpers.py:
```diff
@@ -74,3 +74,21 @@ def convert_to_bell_string(bell: int) -> str:
+
+    start_row = [Bell.from_str(i) for i in start_row_string]
+    if len(start_row) > len(set(start_row)):
+        raise ValueError(f"starting row '{start_row_string}' contains the same bell multiple times")
```
In a similar vein to the other error message, I think this should say actually what bell is duplicated (since that's really helpful for the user).

chrisjfield replied:
Yeah, i thought about that, was just trying to avoid having too many loops but I'll add it in :)

centreboard replied:
I think it's clear enough (the string ought to be at most 16 characters), without needing the logic of which bell it is

> starting row '5321246' contains the same bell multiple times

kneasle commented on wheatley/row_generation/place_notation_generator.py:
```diff
@@ -18,8 +18,8 @@ class PlaceNotationGenerator(RowGenerator):
     def __init__(self, stage: int, method: str, bob: CallDef = None, single: CallDef = None,
-                 start_index: int = 0) -> None:
-        super().__init__(stage)
+                 start_index: int = 0, start_row: str = None) -> None:
```
Shouldn't this be `start_row: Option[str] = None`, since `str` can't be `None`.  I'm honestly surprised that the type checker isn't catching this, but it happens many times across the code (I think grepping for `str = None` should find them all) and it would probably be good to be explicit about it.

centreboard replied:
I recall something about the typechecker allowing this in `__init__` (see the `bob: CallDef = None` ones )

kneasle reviewed (commented):
Looking good!  There are some tweaks to be made, but generally it seems very solid.

One thought I have is that we end up passing `start_row` around **a lot**.  I wonder if it would be cleaner to add a method `set_start_row(new_row: Row)` to `RowGenerator` and then use that after initialisation.  This can then be overriden by `ComplibCompositionGenerator` to throw the error message which will make sure that comps **never** have a custom start row (currently we only check in the CLI main function, but we don't know if `Complib...` is ever constructed elsewhere).  @centreboard, what do you think?

centreboard commented:
> One thought I have is that we end up passing `start_row` around **a lot**. I wonder if it would be cleaner to add a method `set_start_row(new_row: Row)` to `RowGenerator` and then use that after initialisation. This can then be overriden by `ComplibCompositionGenerator` to throw the error message which will make sure that comps **never** have a custom start row (currently we only check in the CLI main function, but we don't know if `Complib...` is ever constructed elsewhere). @centreboard, what do you think?

I prefer the current design of passing it as an argument. So rather than have a method that might throw `ComplibCompositionGenerator` doesn't support it because it's not exposed as a constructor argument.

chrisjfield commented:
> Oh you're running Windows, right?

Yes indeed - guessing you're both on Linux.

> It's git doing auto CRLF conversion on Windows. When it checkouts out files git converts the line endings to CRLF, when you commit it converts back to LF.

Yeah, it's no biggie I'll just persuade VS code to auto-format that workspace to LF to make the local linter happy :)

kneasle commented:
> guessing you're both on Linux.

I am, but I think Matt uses Windows.

So here's what I think we should do:
- Release `v0.6.0` from the master branch as it stands right now.  It has more than enough changes already, and this way we can isolate any bugs more easily.
- We rebase and merge this to prevent more merge conflicts.  I'm happy to do the rebase/combine commits/fix conflicts and merge in one go.
- I still think we should convert to a `Row` in main.py, but I don't think more argument is worthwhile and with Python's lack of immutability I agree that this isn't an option if `Row = List[Bell]`.  However, long term I think it would be a good idea to create an immutable `Row` class that can be extended, parsed, etc. but the operations always return a new `Row` rather than mutating the old one.

centreboard commented:
@chrisjfield Are you happy to merge from master into this branch / rebase it onto master?

You might need to fetch from the upstream `https://github.com/kneasle/wheatley` to get the latest master first.
https://stackoverflow.com/a/15943715

I do use Windows, just mostly using PyCharm instead of VS Code for python.

chrisjfield commented:
Sure,

I'll try to do that this weekend :-) work being crazy!

On Thu, Apr 1, 2021, 22:55 Matthew R Johnson ***@***.***>
wrote:

> @chrisjfield <https://github.com/chrisjfield> Are you happy to merge from
> master into this branch / rebase it onto master?
>
> You might need to fetch from the upstream
> https://github.com/kneasle/wheatley to get the latest master first.
> https://stackoverflow.com/a/15943715
>
> I do use Windows, just mostly using PyCharm instead of VS Code for python.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kneasle/wheatley/pull/159#issuecomment-812193586>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHLTBS5E25NYE7OFR2NYZCLTGTTUXANCNFSM4ZWELHGQ>
> .
>

chrisjfield commented:
Ended up doing it now anyway.

Rebased onto master - seemed ok, just Contributing that conflicted (I used the new version from master), ran the tests locally etc and did a few spot checks.

kneasle commented:
Cool! I'm merging now.
